File: paintbot_lab/paintbot/stencil/perception.py
"""Sprite-label perception: raw SpriteWorld -> PaintState.

Mostly label lookups over the retained scene, as the Nim baseline does it — plus
the three things Paintbot's procgen maps force us to READ off the wire that
beacon compiled in:

  * the ``game teams <count> map <width>x<height>`` marker (episode parameters);
  * the per-team ``endzone <color> <shape> <x0>,<y0> <x1>,<y1>`` markers;
  * the ``walkability map`` sprite's PIXELS (snappy raw-block compressed RGBA;
    alpha>0 = walkable) — decoded once per sprite definition and cached by id.

Team color is initially guessed from the connection slot (slot mod teams) and
locked from the first ``self <color>`` sighting (belief handles the lock).
"""

from __future__ import annotations

import math

import numpy as np

try:
    import cramjam
except ImportError:  # pragma: no cover - cramjam ships in the image
    cramjam = None

from paintbot.stencil.config import AIM_BRADS_TURN, RENDER_SCALE, TEAM_COLORS
from paintbot.stencil.types import Enemy, HeartState, PaintState, Team
from players.player_sdk import SpriteObject, SpriteWorld

#: Carried heart within this distance (px) of us => we're carrying it (the
#: carried banner is CENTERED on its carrier).
_CARRY_DIST = 24.0
#: The white-outlined self soldier sprite pool: 5100 + skin*16 + rot.
_SELF_SPRITE_BASE = 5100
_SOLDIER_ROTATIONS = 16
#: Overhead UI (hp bar, carried-item markers) association radius.
_OVERHEAD_RADIUS = 34.0
#: Pickup sprite labels on the map layer, exact-match.
_ITEM_LABELS = {
    "grenade": "grenade",
    "med kit": "medkit",
    "shield": "shield",
    "plasma arc": "arc",
    "spray can": "arc",
}
#: Sound-ring labels (hearing): audible map-wide, jittered ±20px, team-anonymous.
_SOUND_LABELS = {
    "shot impact": "shot",
    "grenade sound": "grenade",
}
#: Endzone marker shape vocabulary (labels.nim LabelEndzoneShapes). The
#: spectator stream also emits "endzone <color> power <n>" glow overlays;
#: matching the shape token filters those out.
_ENDZONE_SHAPES = ("column", "square", "disc", "corner", "arm")

#: Identity badges (nameplates).
_BADGE_RADIUS = 30.0
_IDENTITY_NAMES = ("alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta", "theta")
_IDENTITY_INDEX = {n: i for i, n in enumerate(_IDENTITY_NAMES)}

#: Decoded walkability masks keyed by (sprite_id, width, height) — a pure decode
#: cache, safe across episodes because a different map defines a different sprite.
_walkability_cache: dict[tuple[int, int, int], np.ndarray] = {}


def _center(world: SpriteWorld, obj: SpriteObject) -> tuple[int, int]:
    """Map-space centre of an object (player wire is 1x map pixels)."""
    sprite = world.sprite_for(obj)
    w = sprite.width if sprite else 0
    h = sprite.height if sprite else 0
    return (round((obj.x + w / 2) / RENDER_SCALE), round((obj.y + h / 2) / RENDER_SCALE))


def _objects_with_label(world: SpriteWorld, label: str) -> list[SpriteObject]:
    out = []
    for obj in world.objects.values():
        sprite = world.sprite_for(obj)
        if sprite is not None and sprite.label == label:
            out.append(obj)
    return out


def _dist(a: tuple[int, int], b: tuple[int, int]) -> float:
    return math.hypot(a[0] - b[0], a[1] - b[1])


# --- Episode parameters (procgen: read, never assume) --------------------------------


def parse_game_params(world: SpriteWorld) -> tuple[int, tuple[int, int]] | None:
    """The ``game teams <count> map <width>x<height>`` marker, if present."""
    for sprite in world.sprites.values():
        label = sprite.label
        if not label.startswith("game teams "):
            continue
        parts = label[len("game teams "):].split(" ")
        # ["<count>", "map", "<width>x<height>"]
        if len(parts) != 3 or parts[1] != "map" or "x" not in parts[2]:
            continue
        try:
            teams = max(2, min(int(parts[0]), 4))
            w, h = (int(v) for v in parts[2].split("x", 1))
        except ValueError:
            continue
        return teams, (w, h)
    return None


def parse_endzones(world: SpriteWorld) -> dict[Team, tuple[str, tuple[int, int, int, int]]]:
    """Per-team endzone markers: color -> (shape, inclusive corner box)."""
    out: dict[Team, tuple[str, tuple[int, int, int, int]]] = {}
    for sprite in world.sprites.values():
        label = sprite.label
        if not label.startswith("endzone "):
            continue
        parts = label[len("endzone "):].split(" ")
        # [<color>, <shape>, "<x0>,<y0>", "<x1>,<y1>"]
        if len(parts) != 4 or parts[0] not in TEAM_COLORS:
            continue
        if parts[1] not in _ENDZONE_SHAPES:
            continue  # e.g. the spectator "power" glow overlay
        try:
            x0, y0 = (int(v) for v in parts[2].split(",", 1))
            x1, y1 = (int(v) for v in parts[3].split(",", 1))
        except ValueError:
            continue
        out[parts[0]] = (parts[1], (x0, y0, x1, y1))
    return out


def decode_walkability(world: SpriteWorld) -> np.ndarray | None:
    """The ``walkability map`` sprite decoded to a bool [H, W] mask, or None.

    RGBA, alpha>0 = walkable, always unscaled 1x map pixels. The compressed
    payload is snappy RAW block format (no stream framing) — cramjam's
    ``decompress_raw`` matches the server's supersnappy output.
    """
    for sprite in world.sprites.values():
        if sprite.label != "walkability map":
            continue
        key = (sprite.sprite_id, sprite.width, sprite.height)
        cached = _walkability_cache.get(key)
        if cached is not None:
            return cached
        if cramjam is None or sprite.width <= 0 or sprite.height <= 0:
            return None
        try:
            raw = bytes(cramjam.snappy.decompress_raw(sprite.data))
        except Exception:
            return None
        if len(raw) != sprite.width * sprite.height * 4:
            return None
        alpha = np.frombuffer(raw, dtype=np.uint8).reshape(
            sprite.height, sprite.width, 4
        )[:, :, 3]
        mask = alpha > 0
        _walkability_cache.clear()  # keep exactly one map resident
        _walkability_cache[key] = mask
        return mask
    return None


# --- Players -------------------------------------------------------------------------


def _find_self(world: SpriteWorld, colors: tuple[Team, ...]):
    """Our self marker across the active colors: (centre, color, facing, aim).

    The self sprite's rotation id (5100 + skin*16 + rot, 16 steps of 16 brads)
    is the only absolute aim readback — coarse but drift-free."""
    for color in colors:
        for facing in ("right", "left"):
            objs = _objects_with_label(world, f"self {color} {facing}")
            if objs:
                obj = objs[0]
                offset = obj.sprite_id - _SELF_SPRITE_BASE
                rot = offset % _SOLDIER_ROTATIONS
                aim = (
                    rot * (AIM_BRADS_TURN // _SOLDIER_ROTATIONS)
                    if offset >= 0
                    else None
                )
                return _center(world, obj), color, facing, aim
    return None, None, None, None


def _identity_badges(world: SpriteWorld, color: Team) -> list[tuple[int, tuple[int, int]]]:
    out: list[tuple[int, tuple[int, int]]] = []
    prefix = f"identity {color} "
    for obj in world.objects.values():
        sprite = world.sprite_for(obj)
        if sprite is None or not sprite.label.startswith(prefix):
            continue
        # The badge tail is "<name>[ shield][ nade] <weapon>" — first token is the name.
        name = sprite.label[len(prefix):].split(" ", 1)[0]
        idx = _IDENTITY_INDEX.get(name)
        if idx is not None:
            out.append((idx, _center(world, obj)))
    return out


def _players_of_color(world: SpriteWorld, color: Team) -> tuple[Enemy, ...]:
    """Visible players of ``color``, with nameplate identity where a badge resolves."""
    players: list[tuple[tuple[int, int], str]] = []
    for facing in ("right", "left"):
        for obj in _objects_with_label(world, f"player {color} {facing}"):
            players.append((_center(world, obj), facing))
    badges = _identity_badges(world, color)
    out: list[Enemy] = []
    for pos, facing in players:
        best_idx: int | None = None
        best_d = _BADGE_RADIUS
        for idx, bpos in badges:
            d = math.hypot(bpos[0] - pos[0], bpos[1] - pos[1])
            if d < best_d:
                best_d = d
                best_idx = idx
        out.append(Enemy(pos=pos, facing=facing, color=color, identity=best_idx))
    return tuple(out)


# --- Items / sounds / shouts / scores ------------------------------------------------


def _visible_items(world: SpriteWorld) -> tuple[tuple[str, tuple[int, int]], ...]:
    out: list[tuple[str, tuple[int, int]]] = []
    for obj in world.objects.values():
        sprite = world.sprite_for(obj)
        if sprite is not None and sprite.label in _ITEM_LABELS:
            out.append((_ITEM_LABELS[sprite.label], _center(world, obj)))
    return tuple(out)


def _heard_impacts(world: SpriteWorld) -> tuple[tuple[str, tuple[int, int]], ...]:
    out: list[tuple[str, tuple[int, int]]] = []
    for obj in world.objects.values():
        sprite = world.sprite_for(obj)
        if sprite is not None and sprite.label in _SOUND_LABELS:
            out.append((_SOUND_LABELS[sprite.label], _center(world, obj)))
    return tuple(out)


def _heard_shouts(
    world: SpriteWorld,
) -> tuple[tuple[str, str, str, tuple[int, int]], ...]:
    """Shout bubbles: (team color, sender name, text, pos). The sender is the
    per-team slot letter (anonymized upstream), split on the LAST ': '."""
    out: list[tuple[str, str, str, tuple[int, int]]] = []
    for obj in world.objects.values():
        sprite = world.sprite_for(obj)
        if sprite is None:
            continue
        label = sprite.label
        if " shout " not in label:
            continue
        team, rest = label.split(" shout ", 1)
        if team not in TEAM_COLORS or ": " not in rest:
            continue
        address, text = rest.rsplit(": ", 1)
        out.append((team, address, text, _center(world, obj)))
    return tuple(out)


def _team_scores(world: SpriteWorld) -> dict[Team, tuple[int, int]]:
    """All teams' (kills, deaths) from the top-center scoreboard chips.

    Labels read ``team score RED <k>/<d>`` per active team, fog-independent."""
    out: dict[Team, tuple[int, int]] = {}
    for obj in world.objects.values():
        sprite = world.sprite_for(obj)
        if sprite is None or not sprite.label.startswith("team score "):
            continue
        parts = sprite.label[len("team score "):].split(" ")
        if len(parts) != 2 or "/" not in parts[1]:
            continue
        color = parts[0].lower()
        if color not in TEAM_COLORS:
            continue
        try:
            kills, deaths = (int(v) for v in parts[1].split("/", 1))
        except ValueError:
            continue
        out[color] = (kills, deaths)
    return out


# --- Overhead markers ----------------------------------------------------------------


def _nearest_marker_assignments(
    actors: list[tuple[int, int]],
    markers: list[tuple[int, int]],
) -> dict[int, int]:
    """Globally assign nearby overhead markers to actors, one-to-one."""
    pairs: list[tuple[float, int, int]] = []
    for actor_i, actor_pos in enumerate(actors):
        for marker_i, marker_pos in enumerate(markers):
            d = _dist(actor_pos, marker_pos)
            if d <= _OVERHEAD_RADIUS:
                pairs.append((d, actor_i, marker_i))
    assigned_actors: set[int] = set()
    assigned_markers: set[int] = set()
    out: dict[int, int] = {}
    for _distance, actor_i, marker_i in sorted(pairs):
        if actor_i in assigned_actors or marker_i in assigned_markers:
            continue
        assigned_actors.add(actor_i)
        assigned_markers.add(marker_i)
        out[actor_i] = marker_i
    return out


def _attach_overhead_state(
    world: SpriteWorld,
    self_xy: tuple[int, int],
    enemies: tuple[Enemy, ...],
    teammates: tuple[Enemy, ...],
):
    """Attach ordinal HP bars and carried markers to every visible player."""
    hp_values: list[int] = []
    hp_positions: list[tuple[int, int]] = []
    carried_positions: dict[str, list[tuple[int, int]]] = {
        "grenade carried": [],
        "shield carried": [],
        "plasma arc carried": [],
        "spray can carried": [],
    }
    for obj in world.objects.values():
        sprite = world.sprite_for(obj)
        if sprite is None:
            continue
        label = sprite.label
        if label.startswith("hp "):
            try:
                lit = int(label.split(" ", 1)[1].split("/", 1)[0])
            except (IndexError, ValueError):
                continue
            hp_values.append(lit)
            hp_positions.append(_center(world, obj))
        elif label in carried_positions:
            carried_positions[label].append(_center(world, obj))

    self_hp: int | None = None
    self_hp_marker: int | None = None
    self_hp_distance = _OVERHEAD_RADIUS
    for marker_i, marker_pos in enumerate(hp_positions):
        distance = _dist(self_xy, marker_pos)
        if distance < self_hp_distance:
            self_hp_distance = distance
            self_hp = hp_values[marker_i]
            self_hp_marker = marker_i

    actors = [*(enemy.pos for enemy in enemies), *(mate.pos for mate in teammates)]
    remaining_hp = [
        (marker_i, marker_pos)
        for marker_i, marker_pos in enumerate(hp_positions)
        if marker_i != self_hp_marker
    ]
    hp_by_actor = {
        actor_i: hp_values[remaining_hp[marker_i][0]]
        for actor_i, marker_i in _nearest_marker_assignments(
            actors,
            [pos for _original_i, pos in remaining_hp],
        ).items()
    }

    self_carried: dict[str, bool] = {}
    carried_by_actor: dict[str, set[int]] = {}
    for label, positions in carried_positions.items():
        self_markers = {
            marker_i
            for marker_i, marker_pos in enumerate(positions)
            if _dist(self_xy, marker_pos) <= _OVERHEAD_RADIUS
        }
        self_carried[label] = bool(self_markers)
        remaining = [
            marker_pos
            for marker_i, marker_pos in enumerate(positions)
            if marker_i not in self_markers
        ]
        carried_by_actor[label] = set(_nearest_marker_assignments(actors, remaining))

    enemy_start = 0
    teammate_start = len(enemies)
    attached_enemies = tuple(
        Enemy(
            pos=enemy.pos,
            facing=enemy.facing,
            color=enemy.color,
            identity=enemy.identity,
            hp_segments=hp_by_actor.get(enemy_start + i),
            shielded=enemy_start + i in carried_by_actor["shield carried"],
        )
        for i, enemy in enumerate(enemies)
    )
    attached_teammates = tuple(
        Enemy(
            pos=mate.pos,
            facing=mate.facing,
            color=mate.color,
            identity=mate.identity,
            hp_segments=hp_by_actor.get(teammate_start + i),
            shielded=teammate_start + i in carried_by_actor["shield carried"],
        )
        for i, mate in enumerate(teammates)
    )
    return (
        self_hp,
        self_carried["grenade carried"],
        self_carried["shield carried"],
        (self_carried["plasma arc carried"] or self_carried["spray can carried"]),
        attached_enemies,
        attached_teammates,
    )


# --- The percept ---------------------------------------------------------------------


def perceive(obs, team: Team, colors: tuple[Team, ...]) -> PaintState:
    """Read one frame's PaintState for our ``team`` among active ``colors``."""
    world = obs.world
    enemy_colors = tuple(c for c in colors if c != team)

    self_xy, self_color, self_facing, observed_aim = _find_self(world, colors)
    if self_color is not None and self_color != team:
        # Slot-dealt guess was wrong; report players relative to the REAL color.
        team = self_color
        enemy_colors = tuple(c for c in colors if c != team)
    ready = self_xy is not None
    fire_ready = len(_objects_with_label(world, "fire icon")) > 0
    enemies: tuple[Enemy, ...] = ()
    for color in enemy_colors:
        enemies = enemies + _players_of_color(world, color)
    teammates = _players_of_color(world, team)

    # Heart bookkeeping per color. A planted heart rests on its pedestal (never
    # fogged); a carried heart is exactly as visible as its carrier. For each
    # color: planted -> stealable/safe; carried-visible -> a live fix; NEITHER
    # sprite -> stolen by a fogged carrier (or retired after elimination —
    # belief.py resolves which via the team scoreboard).
    hearts: dict[Team, HeartState] = {}
    for color in colors:
        planted = _objects_with_label(world, f"{color} flag planted")
        carried = _objects_with_label(world, f"{color} flag")
        pos = _center(world, planted[0]) if planted else None
        carried_pos = _center(world, carried[0]) if carried else None
        hearts[color] = HeartState(
            planted=bool(planted),
            pos=pos if planted else carried_pos,
            carried_pos=carried_pos,
        )

    i_carry: Team | None = None
    if self_xy is not None:
        for color in enemy_colors:
            state = hearts[color]
            if state.carried_pos is not None and _dist(state.carried_pos, self_xy) <= _CARRY_DIST:
                i_carry = color
                break

    visible_items = _visible_items(world)
    heard_impacts = _heard_impacts(world)
    heard_shouts = _heard_shouts(world)
    if self_xy is not None:
        (
            hp_pips,
            have_grenade,
            have_shield,
            have_arc,
            enemies,
            teammates,
        ) = _attach_overhead_state(world, self_xy, enemies, teammates)
        have_arc = have_arc or bool(_objects_with_label(world, "weapon spray"))
    else:
        hp_pips, have_grenade, have_shield, have_arc = None, False, False, False

    params = parse_game_params(world)
    return PaintState(
        ready=ready,
        self_xy=self_xy,
        self_color=self_color,
        self_facing=self_facing,
        observed_aim=observed_aim,
        fire_ready=fire_ready,
        enemies=enemies,
        teammates=teammates,
        hearts=hearts,
        i_carry_heart_of=i_carry,
        game_teams=params[0] if params else None,
        map_size=params[1] if params else None,
        endzones=parse_endzones(world),
        walkability=decode_walkability(world),
        visible_items=visible_items,
        heard_impacts=heard_impacts,
        heard_shouts=heard_shouts,
        hp_pips=hp_pips,
        i_have_grenade=have_grenade,
        i_have_shield=have_shield,
        i_have_arc=have_arc,
        team_scores=_team_scores(world),
    )


__all__ = [
    "decode_walkability",
    "parse_endzones",
    "parse_game_params",
    "perceive",
]

File: paintbot_lab/paintbot/stencil_nim/perception.nim
## Sprite-label perception: retained Sprite-v1 scene to a memoryless PaintState.

import std/[algorithm, math, options, sequtils, sets, strutils, tables]
import config, protocols, types

const
  CarryDistance = 24.0
  SelfSpriteBase = 5100
  PlayerSpriteBase = 100
  SelectedPlayerSpriteBase = 6000
  SoldierSkins = 2
  OverheadRadius = 34.0
  BadgeRadius = 30.0
  IdentityNames = ["alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta", "theta"]

type
  SceneObject = tuple[objectId, spriteId, x, y, width, height: int, label: string]
  SceneIndex = object
    objects: seq[SceneObject]
    byLabel: Table[string, seq[int]]
  ActorMarkerPair = tuple[distance: float, actorIndex, markerIndex: int]

proc center(item: SceneObject): Point =
  (pyRound((item.x.float + item.width.float / 2.0) / RenderScale.float),
   pyRound((item.y.float + item.height.float / 2.0) / RenderScale.float))

proc distance(a, b: Point): float = hypot((a.x - b.x).float, (a.y - b.y).float)

proc objects(client: ProtocolClient): SceneIndex =
  for item in client.spriteObjects:
    let index = result.objects.len
    result.objects.add(item)
    result.byLabel.mgetOrPut(item.label, @[]).add(index)

iterator objectsWithLabel(scene: SceneIndex, label: string): SceneObject =
  if scene.byLabel.hasKey(label):
    for index in scene.byLabel[label]:
      yield scene.objects[index]

proc firstWithLabel(scene: SceneIndex, label: string): Option[SceneObject] =
  if scene.byLabel.hasKey(label) and scene.byLabel[label].len > 0:
    some(scene.objects[scene.byLabel[label][0]])
  else:
    none(SceneObject)

proc hasLabel(scene: SceneIndex, label: string): bool =
  scene.byLabel.hasKey(label) and scene.byLabel[label].len > 0

proc parseGameParams*(client: ProtocolClient): Option[tuple[teams: int, mapSize: Point]] =
  client.gameParams

proc parseEndzones*(client: ProtocolClient): Table[Team, EndzoneMarker] =
  client.endzoneMarkers

proc findSelf(
  scene: SceneIndex, colors: openArray[Team]
): tuple[pos: Option[Point], color: Option[Team], facing: Option[Facing], aim: Option[int]] =
  var exactAim = none(int)
  for item in scene.objects:
    if not item.label.startsWith("own aim "):
      continue
    try:
      let value = parseInt(item.label["own aim ".len .. ^1])
      if value >= 0 and value < AimBradsTurn:
        exactAim = some(value)
    except ValueError:
      discard
  for color in colors:
    for facing in [FacingRight, FacingLeft]:
      let label = "self " & color.teamName & " " &
        (if facing == FacingRight: "right" else: "left")
      let match = scene.firstWithLabel(label)
      if match.isNone:
        continue
      let item = match.get
      return (some(item.center), some(color), some(facing),
        if exactAim.isSome:
          exactAim
        else:
          # The soldier has only 16 visual rotations while the turret uses all
          # 256 integer-brad headings. Keep this as a compatibility fallback.
          let spriteId = item.spriteId
          if spriteId >= SelfSpriteBase:
            some(((spriteId - SelfSpriteBase) mod ObservedHeadingSteps) *
              (AimBradsTurn div ObservedHeadingSteps))
          else:
            none(int))

proc identityBadges(
  scene: SceneIndex, color: Team
): seq[tuple[identity: int, pos: Point]] =
  let prefix = "identity " & color.teamName & " "
  for item in scene.objects:
    if not item.label.startsWith(prefix):
      continue
    let name = item.label[prefix.len .. ^1].split(' ', 1)[0]
    let identity = IdentityNames.find(name)
    if identity >= 0:
      result.add((identity, item.center))

proc playersOfColor(scene: SceneIndex, color: Team): seq[Enemy] =
  let badges = identityBadges(scene, color)
  for facing in [FacingRight, FacingLeft]:
    let label = "player " & color.teamName & " " &
      (if facing == FacingRight: "right" else: "left")
    for item in objectsWithLabel(scene, label):
      let pos = item.center
      let aimBrads =
        if item.spriteId >= PlayerSpriteBase and
            item.spriteId < PlayerSpriteBase +
              SoldierSkins * TeamColors.len * ObservedHeadingSteps:
          some(((item.spriteId - PlayerSpriteBase) mod ObservedHeadingSteps) *
            (AimBradsTurn div ObservedHeadingSteps))
        elif item.spriteId >= SelectedPlayerSpriteBase and
            item.spriteId < SelectedPlayerSpriteBase +
              SoldierSkins * TeamColors.len * ObservedHeadingSteps:
          some(((item.spriteId - SelectedPlayerSpriteBase) mod ObservedHeadingSteps) *
            (AimBradsTurn div ObservedHeadingSteps))
        else:
          none(int)
      var
        bestIdentity = none(int)
        bestDistance = BadgeRadius
      for badge in badges:
        let d = distance(pos, badge.pos)
        if d < bestDistance:
          bestDistance = d
          bestIdentity = some(badge.identity)
      result.add(Enemy(pos: pos, facing: facing, aimBrads: aimBrads, color: color,
        identity: bestIdentity, hpSegments: none(int)))

proc visibleItems(scene: SceneIndex): seq[VisibleItem] =
  for item in scene.objects:
    let kind = case item.label
      of "grenade": "grenade"
      of "med kit": "medkit"
      of "shield": "shield"
      of "plasma arc", "spray can": "arc"
      else: ""
    if kind.len > 0:
      result.add((kind, item.center))

proc heardImpacts(scene: SceneIndex): seq[HeardSound] =
  for item in scene.objects:
    let kind = case item.label
      of "shot impact": "shot"
      of "grenade sound": "grenade"
      else: ""
    if kind.len > 0:
      result.add((kind, item.center))

proc heardShouts(scene: SceneIndex): seq[HeardShout] =
  for item in scene.objects:
    let marker = item.label.find(" shout ")
    if marker < 0:
      continue
    let team = item.label[0 ..< marker]
    if parseTeam(team).isNone:
      continue
    let rest = item.label[marker + " shout ".len .. ^1]
    let separator = rest.rfind(": ")
    if separator < 0:
      continue
    result.add((team, rest[0 ..< separator], rest[separator + 2 .. ^1], item.center))

proc teamScores(scene: SceneIndex): Table[Team, tuple[kills, deaths: int]] =
  const Prefix = "team score "
  for item in scene.objects:
    if not item.label.startsWith(Prefix):
      continue
    let parts = item.label[Prefix.len .. ^1].split(' ')
    if parts.len != 2:
      continue
    let team = parseTeam(parts[0].toLowerAscii)
    let score = parts[1].split('/', 1)
    if team.isNone or score.len != 2:
      continue
    try:
      result[team.get] = (parseInt(score[0]), parseInt(score[1]))
    except ValueError:
      discard

proc markerAssignments(
  actors, markers: openArray[Point]
): Table[int, int] =
  var pairs: seq[ActorMarkerPair]
  for actorIndex, actor in actors:
    for markerIndex, marker in markers:
      let d = distance(actor, marker)
      if d <= OverheadRadius:
        pairs.add((d, actorIndex, markerIndex))
  pairs.sort(proc(a, b: ActorMarkerPair): int = cmp(a.distance, b.distance))
  var
    assignedActors: HashSet[int]
    assignedMarkers: HashSet[int]
  for pair in pairs:
    if pair.actorIndex in assignedActors or pair.markerIndex in assignedMarkers:
      continue
    assignedActors.incl(pair.actorIndex)
    assignedMarkers.incl(pair.markerIndex)
    result[pair.actorIndex] = pair.markerIndex

proc attachOverheadState(
  scene: SceneIndex, selfXy: Point,
  enemies, teammates: var seq[Enemy]
): tuple[hp: Option[int], grenade, shield, arc: bool] =
  var
    hpValues: seq[int]
    hpPositions: seq[Point]
    carried = initTable[string, seq[Point]]()
  for label in ["grenade carried", "shield carried", "plasma arc carried", "spray can carried"]:
    carried[label] = @[]
  for item in scene.objects:
    if item.label.startsWith("hp "):
      let fields = item.label.split({' ', '/'})
      if fields.len >= 2:
        try:
          hpValues.add(parseInt(fields[1]))
          hpPositions.add(item.center)
        except ValueError:
          discard
    elif carried.hasKey(item.label):
      carried[item.label].add(item.center)

  var
    selfHpMarker = -1
    selfHpDistance = OverheadRadius
  result.hp = none(int)
  for markerIndex, markerPos in hpPositions:
    let d = distance(selfXy, markerPos)
    if d < selfHpDistance:
      selfHpDistance = d
      selfHpMarker = markerIndex
      result.hp = some(hpValues[markerIndex])

  let actors = enemies.mapIt(it.pos) & teammates.mapIt(it.pos)
  var remainingHpIndices: seq[int]
  var remainingHpPositions: seq[Point]
  for markerIndex, markerPos in hpPositions:
    if markerIndex != selfHpMarker:
      remainingHpIndices.add(markerIndex)
      remainingHpPositions.add(markerPos)
  for actorIndex, markerIndex in markerAssignments(actors, remainingHpPositions):
    let hp = hpValues[remainingHpIndices[markerIndex]]
    if actorIndex < enemies.len:
      enemies[actorIndex].hpSegments = some(hp)
    else:
      teammates[actorIndex - enemies.len].hpSegments = some(hp)

  for label, positions in carried:
    var
      selfMarkers: HashSet[int]
      remaining: seq[Point]
    for markerIndex, markerPos in positions:
      if distance(selfXy, markerPos) <= OverheadRadius:
        selfMarkers.incl(markerIndex)
      else:
        remaining.add(markerPos)
    let hasSelf = selfMarkers.len > 0
    case label
    of "grenade carried": result.grenade = hasSelf
    of "shield carried": result.shield = hasSelf
    of "plasma arc carried", "spray can carried": result.arc = result.arc or hasSelf
    else: discard
    if label == "shield carried":
      for actorIndex in markerAssignments(actors, remaining).keys:
        if actorIndex < enemies.len:
          enemies[actorIndex].shielded = true
        else:
          teammates[actorIndex - enemies.len].shielded = true

proc perceive*(
  client: ProtocolClient, team: Team, colors: openArray[Team],
  includeWalkability = true
): PaintState =
  let scene = client.objects
  let self = findSelf(scene, colors)
  var actualTeam = team
  if self.color.isSome:
    actualTeam = self.color.get

  result.selfXy = self.pos
  result.selfColor = self.color
  result.selfFacing = self.facing
  result.observedAim = self.aim
  result.ready = self.pos.isSome
  result.fireReady = scene.hasLabel("fire icon")

  for color in colors:
    if color == actualTeam:
      result.teammates.add(playersOfColor(scene, color))
    else:
      result.enemies.add(playersOfColor(scene, color))
  for color in colors:
    let
      planted = scene.firstWithLabel(color.teamName & " flag planted")
      carried = scene.firstWithLabel(color.teamName & " flag")
      plantedPos = if planted.isSome: some(planted.get.center) else: none(Point)
      carriedPos = if carried.isSome: some(carried.get.center) else: none(Point)
    result.hearts[color] = HeartState(
      planted: planted.isSome,
      pos: if plantedPos.isSome: plantedPos else: carriedPos,
      carriedPos: carriedPos)

  result.iCarryHeartOf = none(Team)
  if self.pos.isSome:
    for color in colors:
      if color == actualTeam:
        continue
      let heart = result.hearts[color]
      if heart.carriedPos.isSome and distance(heart.carriedPos.get, self.pos.get) <= CarryDistance:
        result.iCarryHeartOf = some(color)
        break
  result.visibleItems = visibleItems(scene)
  result.heardImpacts = heardImpacts(scene)
  result.heardShouts = heardShouts(scene)
  if self.pos.isSome:
    let overhead = attachOverheadState(scene, self.pos.get, result.enemies, result.teammates)
    result.hpPips = overhead.hp
    result.iHaveGrenade = overhead.grenade
    result.iHaveShield = overhead.shield
    result.iHaveArc = overhead.arc or scene.hasLabel("weapon spray")
  else:
    result.hpPips = none(int)
  let params = parseGameParams(client)
  if params.isSome:
    result.gameTeams = some(params.get.teams)
    result.mapSize = some(params.get.mapSize)
  else:
    result.gameTeams = none(int)
    result.mapSize = none(Point)
  result.endzones = parseEndzones(client)
  if includeWalkability and client.walkabilityReady:
    result.walkability = client.walkabilityMask
    result.walkabilityWidth = client.walkabilityWidth
    result.walkabilityHeight = client.walkabilityHeight
  result.teamScores = teamScores(scene)
